Generate unique id for tensor storage object

shengfukevin opened the pull request:
Summary:
PyTorch execution trace records tensor storage data in the trace. The tensor storage data includes storage id, offset, number of elements, and number of byte for each element. PARAM et-replay uses this information to allocate/free the tensors.
However, the current implementation of generating tensor storage id does not guarantee it is unique. ExecutionTraceObserver maintains a lookup table to map the memory address of the tensor storage object to an unique id. If a new memory address is found, it will be put into that hash table and associate it to a new id.
This implementation does not guarantee the storage object is unique since the memory that the address points to may be released and then re-allocated to a different tensor storage object.
To fix it, a static global id is used to generate the unique id for storage object. The global id is an atomic variable and initialized to 0. Then when a new storage object is created, its id is set to the global id, and then bump up the global id.

Test Plan: buck2 run mode/opt caffe2/test:torch -- -r test_torch.TestTorchDeviceTypeCPU.test_storage_*

Differential Revision: D75045446

shengfukevin commented:
@pytorchbot label "topic: not user facing"

Skylion007 commented on c10/core/StorageImpl.h:
```diff
@@ -93,6 +93,20 @@ struct C10_API StorageImpl : public c10::intrusive_ptr_target {
   ~StorageImpl() override = default;
 
+  // NOLINT
+  static std::atomic<size_t> g_id;
```
If all we care about is uniqueness, why not just use a UUID approach which doesn't require any static atomic locks?

shengfukevin replied:
Thanks for your comments. 
I prefer the atomic variable than UUID for the following reasons:
1. the size of size_t is 64 bits, UUID is 128 bits, there are large number of tensor storage in the trace, it will increase the size of the trace.
2. size_t is a intrinsic data type, it is much easier to handle in execution trace related tools than UUID. These  tools already assumes storage id is a unsigned long, change it to UUID requires lots of work.
3. while UUIDs are extremely unlikely to collide, they do not guarantee absolute uniqueness. It might become a hard to debug bug if it happens.

Skylion007 replied:
And we don't expect the atomic locks of g_id to become a significant performance bottleneck?

shengfukevin replied:
GPU memory address can be reused for different tensors. For example, I have tensor_a, and have tensor storage pointing to GPU address address_1, then call op1(tensor_a), and tensor_a got freed. Later I got tensor_b and its storage also pointing to GPYU address address_1, then call op2(tensor_b). tensor_a and tensor_b are different tensors, they just happened to share the same memory address. However it ET, we use tensor_storage_id to identify the tensor, it ends up with the same tensor passed into both op1 and op2.

Atomic variable is not a mutex lock, it is just guaranteed certain memory order is enforced across threads. In PyTorch, usually only one thread for FWD pass, and one thread for BWD pass, I do not think it will be a performance bottleneck.

shengfukevin replied:
This is from Metamate 

https://www.internalfb.com/metamate/1340761046994971

shengfukevin replied:
I am going to write a stress test to test the overhead of atomic variable in tensor storage.

shengfukevin replied:
I added two stress tests: 

1. test the performance of creating a new tensor storage object with multiple threads, the following are the time for ten runs:

With storage id change | 81 | 83 | 81 | 81 | 85 | 80 | 80 | 82 | 83 | 82 | avg | 81.8

Without storage id change | 82 | 84 | 83 | 81 | 81 | 82 | 83 | 82 | 80 | 81 | avg | 81.9

The difference is just noise. Since get_id is a new function, the performance of the existing models won't be affected.

2. test the uniqueness of the storage tensor id with multiple threads, it passed.

@Skylion007, please take a look.

eellison reviewed (changes requested):
This doesnt seem like the right solution. We should not add another atomic to the hot path of storage allocation. 

If you need to distinguish ids, you might try weak_ptr, or python weakref instead, and track current lifetime of storage. Should be done on the user side.

shengfukevin commented:
> This doesnt seem like the right solution. We should not add another atomic to the hot path of storage allocation.
The only overhead in hot path is to initialize the atomic id to 0, get_id() is not in hot path. From my stress test, it does not affect the performance.

> 
> If you need to distinguish ids, you might try weak_ptr, or python weakref instead, and track current lifetime of storage. Should be done on the user side.

Do you mean Storage:getWeakStorageImpl? It is a cpu memory pointer, also could be reused. How do I track the lifetime of the storage in pytorch execution trace observer?

shengfukevin committed:
Generate unique id for tensor storage object (#153921)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/153921

PyTorch execution trace records tensor storage data in the trace. The tensor storage data includes storage id, offset, number of elements, and number of byte for each element. PARAM et-replay uses this information to allocate/free the tensors.
However, the current implementation of generating tensor storage id does not guarantee it is unique. ExecutionTraceObserver maintains a lookup table to map the memory address of the tensor storage object to an unique id. If a new memory address is found, it will be put into that hash table and associate it to a new id.
This implementation does not guarantee the storage object is unique since the memory that the address points to may be released and then re-allocated to a different tensor storage object.
To fix it, a static global id is used to generate the unique id for storage object. The global id is an atomic variable and initialized to 0. When a new storage object is created, its id is simply set to 0, so it won't affect the performance of creating a new storage object. When get_id method is called on Storage object, we check if its id is zero or not, if it is zero, set it to a new id fetched from global id in a thread safe way; otherwise, just return id. Since get_id is a new method, it won't affect the performance of creating the new Storage object. The only cost to the existing code is the extra memory of sizeof(std::atomic<size_t>).
This implementation guaranteed the id is unique, but does not guarantee the id following the order that storage object is created.

Test Plan:
buck2 run mode/opt caffe2/c10/test:core_tests -- --gtest_filter=storage_id_test.performance

buck2 run mode/opt caffe2/c10/test:core_tests -- --gtest_filter=storage_id_test.uniqueness

Differential Revision: D75045446

ngimel reviewed (approved):
This looks fine to me, no more atomics in the hot path. Is `get_id` on the hot path for you?

ngimel commented:
Globally, I still don't understand how storage_id + offset + numel + element_size can be used to uniquely identify the tensor, I can have multiple aliases and views all sharing this same identifying metadata. How are we using this in ExecutionTrace?

shengfukevin commented:
> This looks fine to me, no more atomics in the hot path. Is `get_id` on the hot path for you?

Yes, however, ExecutionTraceObserver is using a global mutex to get the ID (https://github.com/pytorch/pytorch/blob/main/torch/csrc/profiler/standalone/execution_trace_observer.cpp#L378), which is even slower than using atomic variable.

shengfukevin commented:
@pytorchbot merge

shengfukevin commented:
> Globally, I still don't understand how storage_id + offset + numel + element_size can be used to uniquely identify the tensor, I can have multiple aliases and views all sharing this same identifying metadata. How are we using this in ExecutionTrace?

ET also records shapes, strides, data type, we combine these with storage id to identify the tensor.

eellison commented:
Just for posterity - it's pretty easy to do something like this on the user side. https://gist.github.com/eellison/0eff536cc90b27983eb458d7a69538c7

I think this is better to do on the user side than add overhead to core.

shengfukevin commented:
> Just for posterity - it's pretty easy to do something like this on the user side. https://gist.github.com/eellison/0eff536cc90b27983eb458d7a69538c7. Since ET Observer is already accessing the tensors, you don't need torchdispatchmode, that's just how it is implemented here.
> 
> I think this is better to do on the user side than add overhead to core.

ET is part of PyTorch profiler, which also collects performance data (time) for each function call, the overhead from the solution Elias proposed might skew the performance data of the profiler.

eellison commented:
@shengfukevin only way to get numbers is to try.

shengfukevin commented:
@eellison, I implemented the custom deleter function solution based on your suggestion. This is the PR https://github.com/pytorch/pytorch/pull/154859. Please take a look if it looks good to you. I will try to get some performance data to measure the overhead. Thanks

eellison reviewed (commented):
replaced with https://github.com/pytorch/pytorch/pull/154859
